### PR TITLE
Alerting: Consume k8s Time Intervals API

### PR DESCRIPTION
**Pull request by tomratcliffe**
**What is this feature?**
(Conditionally) Consumes k8s-style APIs for managing mute timings for the Grafana alertmanager.

**Why do we need this feature?**
As part of the move to provide RBAC for managing alertmanager config entities, we need a more granular approach for the built in alertmanager.

This PR will use the kubernetes API if
* The alertmanager used is Grafana's
* The `alertingApiServer` feature toggle is enabled

**Who is this feature for?**

Alerting developers (for now). Will eventually help enable RBAC in more areas of the platform

### CODE REVIEW

File: public/app/features/alerting/unified/state/actions.ts [hunk 1]
```diff
@@ -608,31 +608,29 @@ export const deleteMuteTimingAction = (alertManagerSourceName: string, muteTimin
...
-              ...config,
-              alertmanager_config: {
-                ...configWithoutMuteTimings,
-                route: config.alertmanager_config.route
-                  ? removeMuteTimingFromRoute(muteTimingName, config.alertmanager_config?.route)
-                  : undefined,
-                ...time_intervals_without_mute_to_save,
-              },
+    const { mute_time_intervals: _, ...configWithoutMuteTimings } = config?.alertmanager_config ?? {};
+    return withAppEvents(
```
Comment (tomratcliffe) [inline public/app/features/alerting/unified/state/actions.ts:612]: Change here was just to return the dispatched action so we can consistently chain off it inside useMuteTimings

File: public/app/plugins/datasource/alertmanager/types.ts [hunk 2]
```diff
@@ -330,7 +331,8 @@ export interface TimeInterval {
 export type MuteTimeInterval = {
   name: string;
   time_intervals: TimeInterval[];
-  provenance?: string;
+  provisioned?: boolean;
+  metadata?: ReadNamespacedTimeIntervalApiResponse['metadata'];
```
Comment (tomratcliffe) [inline public/app/plugins/datasource/alertmanager/types.ts:335]: (I should possibly make a separate type that has this property on it instead of adding to this one here, as it's now directly from the datasource)

**Comment by tomratcliffe** [comment]
(I enabled the feature toggle on the ephemeral instance but there's an error from the API - will ask about it later today :+1:)

File: public/app/features/alerting/unified/MuteTimings.test.tsx [hunk 3]
```diff
@@ -143,169 +155,135 @@ const defaultConfigWithBothTimeIntervalsField: AlertManagerCortexConfig = {
   template_files: {},
 };
 
-const resetMocks = () => {
-  jest.resetAllMocks();
+const expectToHaveRedirectedToTable = async () => expect(await screen.findByText(indexPageText)).toBeInTheDocument();
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/MuteTimings.test.tsx:158]: nit: I would change this variable name as its not a table Something like `expectToHaveRedirectedToRightRoute`

File: public/app/features/alerting/unified/MuteTimings.test.tsx [hunk 4]
```diff
@@ -143,169 +155,135 @@ const defaultConfigWithBothTimeIntervalsField: AlertManagerCortexConfig = {
...
+
+const fillOutForm = async ({
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/MuteTimings.test.tsx:160]: Nice!❤️

File: public/app/features/alerting/unified/NotificationPolicies.tsx [hunk 5]
```diff
@@ -65,6 +65,7 @@ const AmRoutes = () => {
 
   const { selectedAlertmanager, hasConfigurationAPI, isGrafanaAlertmanager } = useAlertmanager();
   const { getRouteGroupsMap } = useRouteGroupsMatcher();
+  const { data: muteTimings } = useMuteTimings({ alertmanager: selectedAlertmanager ?? '' });
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/NotificationPolicies.tsx:68]: maybe we should default to `GRAFANA_RULES_SOURCE_NAME` instead of an empty string?

File: public/app/features/alerting/unified/components/DynamicTable.tsx [hunk 6]
```diff
@@ -108,7 +110,7 @@ export const DynamicTable = <T extends object>({
           {renderPrefixHeader && renderPrefixHeader()}
           {isExpandable && <div className={styles.cell} />}
           {cols.map((col) => (
-            <div className={styles.cell} key={col.id}>
+            <div className={styles.cell} key={col.id} style={{ justifyContent: col.alignColumn || 'initial' }}>
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/DynamicTable.tsx:113]: why are we using style here? instead of including this style in the classname?

File: public/app/features/alerting/unified/components/mute-timings/useExportMuteTimingsDrawer.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+export const useExportMuteTimingsDrawer = (): ExportProps => {
+  const [muteTimingName, setMuteTimingName] = useState<string | typeof ALL_MUTE_TIMINGS | null>(null);
+  const [isExportDrawerOpen, toggleShowExportDrawer] = useToggle(false);
+
+  const handleClose = useCallback(() => {
+    setMuteTimingName(null);
+    toggleShowExportDrawer(false);
+  }, [toggleShowExportDrawer]);
+
+  const handleOpen = (receiverName: string | typeof ALL_MUTE_TIMINGS) => {
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/mute-timings/useExportMuteTimingsDrawer.tsx:19]: `receivername` should be `muteTimingName` I guess?

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM! 🚀 I added a non blocking comments

File: public/app/features/alerting/unified/MuteTimings.tsx [hunk 8]
```diff
@@ -1,80 +1,53 @@
...
 import { AlertmanagerPageWrapper } from './components/AlertingPageWrapper';
 import MuteTimingForm from './components/mute-timings/MuteTimingForm';
-import { useAlertmanagerConfig } from './hooks/useAlertmanagerConfig';
 import { useAlertmanager } from './state/AlertmanagerContext';
 
-const MuteTimings = () => {
+const EditTimingRoute = () => {
   const [queryParams] = useQueryParams();
   const { selectedAlertmanager } = useAlertmanager();
-  const { currentData, isLoading, error } = useAlertmanagerConfig(selectedAlertmanager, {
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/MuteTimings.tsx:17]: Agree with separate refactor, though I don't believe many users will have so many mute timings that this might become an issue any time soon.

File: public/app/features/alerting/unified/NotificationPolicies.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/NotificationPolicies.tsx:68]: Is adding support for pagination a prerequisite to get the totals from the API?

[the remark this answers is not in the document]
Comment (tomratcliffe) [reply public/app/features/alerting/unified/NotificationPolicies.tsx:68]: Not sure actually - but right now, as the limit doesn't do anything, I have to query for everything anyway and check the length. If `limit` was functional, then I could ask for 1 item and check the metadata - but I don't think it'll matter too much for now?

File: public/app/features/alerting/unified/MuteTimings.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomratcliffe) [reply public/app/features/alerting/unified/MuteTimings.tsx:17]: Yep, a refactor is more just for a cleaner separation of the logic in our routes etc :+1:

File: public/app/features/alerting/unified/NotificationPolicies.tsx [hunk 9]
```diff
@@ -207,9 +208,8 @@ const AmRoutes = () => {
   if (!selectedAlertmanager) {
     return null;
   }
-  const time_intervals = result?.alertmanager_config ? mergeTimeIntervals(result?.alertmanager_config) : [];
 
-  const numberOfMuteTimings = time_intervals.length;
+  const numberOfMuteTimings = muteTimings?.length || 0;
```
[the remark this answers is not in the document]
Comment (tomratcliffe) [reply public/app/features/alerting/unified/NotificationPolicies.tsx:212]: I agree - however, the reason that this one is potentially undefined is because it's driven by RTKQ-style hooks, and `muteTimings` is potentially undefined if the API request hasn't yet completed. What this should/could do is show a loading spinner for the count badge while its loading. That could come along with a refactor to refactor the routes so this logic isn't even in here?

I considered returning a loading component if the timings were still fetching, but that would mean the entire policies page would be hidden until a small part of the UI had completed loading

File: public/app/plugins/datasource/alertmanager/types.ts [hunk 10]
```diff
@@ -330,7 +330,7 @@ export interface TimeInterval {
 export type MuteTimeInterval = {
   name: string;
   time_intervals: TimeInterval[];
-  provenance?: string;
```
Comment (konrad147) [inline public/app/plugins/datasource/alertmanager/types.ts:333]: Why the change? Isn't it the API response type?

**Comment by tomratcliffe** [comment]
@gillesdemey do you have any thoughts on this? And @konrad147 I've addressed your comments, what do you think? Are we able to merge this in (as its also behind a feature toggle still)?

File: public/app/features/alerting/unified/components/mute-timings/useMuteTimings.tsx [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+
+  useEffect(() => {
+    if (useK8sApi) {
+      const namespace = getNamespace();
+      getGrafanaTimeIntervals({ namespace });
+    } else {
+      getAlertmanagerTimeIntervals(alertmanager);
+    }
+  }, [alertmanager, getAlertmanagerTimeIntervals, getGrafanaTimeIntervals, useK8sApi]);
+  return useK8sApi ? intervalsResponse : configApiResponse;
```
[the remark this answers is not in the document]
Comment (tomratcliffe) [reply public/app/features/alerting/unified/components/mute-timings/useMuteTimings.tsx:129]: (as per our earlier chat, this does in fact include the loading/error/all other RTKQ properties, as its the whole response object)

**Review by konrad147 — APPROVED** [review]
LGTM!

File: public/app/plugins/datasource/alertmanager/types.ts (hunk 10 above)
Comment (gillesdemey) [reply public/app/plugins/datasource/alertmanager/types.ts:333]: I _think_ this is the "UI" data model and not the DTO.

**Review by gillesdemey — APPROVED** [review]
just a few nits; feel free to ignore those! Great work! 🚀
